### PR TITLE
Introduce evaluation option of loss function to SSD application

### PR DESCRIPTION
**Pull request by vuiseng9**
Backward compatible changes. Introduce a config key "loss_inference" to specify test with loss evaluation or the default mAP evaluation. It is disabled by default. This patch support issue #131

### CODE REVIEW

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
@vuiseng9 please fix pylint and then we'll merge this.

**Comment by vuiseng9** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

**Comment by vshampor** [comment]
Jenkins please retry a build

File: examples/object_detection/configs/ssd300_vgg_voc_int8.json [hunk 1]
```diff
@@ -34,7 +34,8 @@
     "steps": [8, 16, 32, 64, 100, 300],
     "aspect_ratios": [[2], [2, 3], [2, 3], [2, 3], [2], [2]],
     "flip": true,
-    "top_k": 200
+    "top_k": 200,
+    "loss_inference": false
```
Comment (vshampor) [inline examples/object_detection/configs/ssd300_vgg_voc_int8.json:38]: We actually use this config file for checkpoint validation, so we require the proper metric calculation.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: examples/object_detection/configs/ssd300_vgg_voc_int8.json (hunk 1 above)
Comment (vuiseng9) [reply examples/object_detection/configs/ssd300_vgg_voc_int8.json:38]: @vshampor When the key is enabled, the training loss is evaluated instead of the mean AP. Disabling the key has the exact behavior as previous implementation. Can you elaborate the metric calculation you refer to?

Comment (ljaljushkin) [reply examples/object_detection/configs/ssd300_vgg_voc_int8.json:38]: I guess @vshampor meant [this](https://github.com/openvinotoolkit/nncf/blob/develop/examples/object_detection/main.py#L189) metric calculation.
It's used in the nightly checkpoint validation defined in test_sota_checkpoints.py. 

But it shouldn't be affected, as you say: the introduced option keeps the previous behavior with calculation actual mAP metric.

File: examples/object_detection/main.py [hunk 2]
```diff
@@ -181,9 +181,14 @@ def criterion_fn(model_outputs, target, criterion):
...
-            if config.metrics_dump is not None:
-                write_metrics(mAp, config.metrics_dump)
+            if config['ssd_params'].get('loss_inference', False):
+                model_loss = test_net(net, config.device, test_data_loader, distributed=config.distributed,
+                                      loss_inference=True, criterion=criterion)
+                logger.info("Final model loss: {:.3f}".format(model_loss))
+            else:
+                mAp = test_net(net, config.device, test_data_loader, distributed=config.distributed)
+                if config.metrics_dump is not None:
+                    write_metrics(mAp, config.metrics_dump)
```
Comment (vshampor) [inline examples/object_detection/main.py:191]: I think that we won't fall into this branch anymore with the config you've changed. We need that config to execute this code branch so that the example script, as run from within the CI validation run, produces an output file with metrics which is then parsed for a test report.

If you modified the config file to give an example of the option's usage, I suggest that you just create another config file in addition to the original one.

**Review by vuiseng9 — COMMENTED** [review]
```config['ssd_params'].get('loss_inference', False)```. When the key is not provided, the app would go as default to mean AP evaluation block. so, the original config and CI validation would still works. does this make sense? or am I disconnected at some points?

File: examples/object_detection/main.py (hunk 2 above)
Comment (vuiseng9) [reply examples/object_detection/main.py:191]: ```config['ssd_params'].get('loss_inference', False)```. When the key is not provided, the app would go as default to mean AP evaluation block. so, the original config and CI validation would still works. does this make sense? or am I disconnected at some points?

Comment (vuiseng9) [reply examples/object_detection/main.py:191]: The config modification is not meant to be as sample, just laying out the key for reference, ie setting it to false so it is backward compatible. Skipping the key is also handled in the main app.

**Review by vshampor — APPROVED** [review]
